pkg/messageprocessors/devicerepository/devicerepository.go

```diff
@@ -0,0 +1,124 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package devicerepository
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/bluele/gcache"
+	"go.thethings.network/lorawan-stack/v3/pkg/messageprocessors"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+	"golang.org/x/sync/singleflight"
+	"google.golang.org/grpc"
+)
+
+const (
+	downlinkEncoder = "downlinkEncoder"
+	downlinkDecoder = "downlinkDecoder"
+	uplinkDecoder   = "uplinkDecoder"
+)
+
+// GetClientConnFunc is a function for retrieving a *grpc.ClientConn to the device repository peer.
+type GetClientConnFunc func() (*grpc.ClientConn, error)
+
+// GetCallOptionsFunc is a function for retrieving grpc.CallOptions for the device repository grpc client.
+type GetCallOptionsFunc func() []grpc.CallOption
+
+type host struct {
+	ctx context.Context
+
+	getClientConn  GetClientConnFunc
+	getCallOptions GetCallOptionsFunc
+
+	singleflight singleflight.Group
+	cache        gcache.Cache
+
+	processor messageprocessors.PayloadProcessor
+}
+
+// New creates a new PayloadEncodeDecoder that retrieves codecs from the Device Repository
+// and uses an underlying PayloadEncodeDecoder to execute them.
+func New(processor messageprocessors.PayloadProcessor, getClientConn GetClientConnFunc, getCallOptions GetCallOptionsFunc) messageprocessors.PayloadEncodeDecoder {
+	return &host{
+		getClientConn:  getClientConn,
+		getCallOptions: getCallOptions,
+
+		processor: processor,
+
+		cache: gcache.New(200).LFU().Expiration(time.Hour).Build(),
+	}
+}
+
+func key(version *ttnpb.EndDeviceVersionIdentifiers, which string) string {
+	return fmt.Sprintf("%s:%s:%s:%s:%v", version.BrandID, version.ModelID, version.FirmwareVersion, version.BandID, which)
+}
+
+func (h *host) retrieve(key string, f func(cc *grpc.ClientConn, opts ...grpc.CallOption) (interface{}, error)) (*ttnpb.MessagePayloadFormatter, error) {
+	if cached, err := h.cache.Get(key); err == nil {
```

Comment:
Where is this `Set` ?
Also, consider saving the errors in the cache as well - you may lower the expiration time for this, of course.